### DISCUSSION TITLE
Any reason why $persist does not have a delete?

### DISCUSSION THREAD

**Discussion by lmajano** [alpinejs/alpine#1972] [4. General]
Is there any reason why the $persist plugin doesn't sport a `delete()` capability?

**Comment by SimoTod** [comment]
I don't think there are specific reasons. Do you have a specific use case and a syntax in mind?

**Comment by lmajano** [comment]
There are certain situations where we don't need to mutate or set a value in local storage, but we want to be able to delete it completely.  It could be due to migrations to a new version of the software and we want to make sure certain values are not there. I know I can go directly to local storage. But if we are going to be providing an abstraction to persistence in local storage. Might as well provide a full abstraction.  This also helps when doing unit tests as I can mock the persistence library methods instead of the local storage objects.

I would suggest that since we already provide a way to do persistence (set) and set with alias, why not include an `exists()` and `delete()` as well to complete the life-cycle and abstractions.

**Reply by SimoTod** [reply]
Yeah, i get that. I was interested in the syntax. Currently you have `$persist(default)` to retrieve a value if it exists or use default.
It also relies on the property name to decide the storage key. Currently I don't see how it could work without massive changes. 
If you introduce something like `$persist(default).fresh()`, it just set the default every time your user visit the page so it will just be a normal property.
In theory you can also use `localstorage.clear()` from vanilla js to remove all your persisted value for that site.

**Reply by dfallman** [reply]
I'd like to bump this one. We've run into a problem with this. We're using $persist to allow users to store a bunch of app settings. One such setting is a set of widgets that the user can organize:

```
showWidgets: $persist([
   { id: 'bananas', name: 'Bananas', icon: '...', order: 0, expanded: true, visible: true, minimized: false, colSpan: 1 },
   { id: 'pears', name: 'Pears', icon: '...', order: 1, expanded: false, visible: true, minimized: false, colSpan: 2 },
   { id: 'apples', name: 'Apples', icon: '...', order: 2, expanded: false, visible: true, minimized: false, colSpan: 1 }
   { ... }
   ]),
```

The problem with this is that if we want to add more widgets in a future release, we need to be able to delete/update `showWidgets`. If we don't and add items to the above, users who already have these settings saved will get an error. 

(I obviously understand that there are other ways to save settings, but) in the specific case of `$persist`, there is then no way of deleting and/or defaulting a single variable? 

In a situation like this, we obviously don't want to delete all the other variables, just the ones that have changed. What are our best options?

**Reply by ekwoka** [reply]
Well, the questions quickly becomes: How do you tell who has the old data and who has the new?

How do you make it delete the old invalid ones and not the new ones?

And if you can figure that out in a way that tells persist to do it, you're basically just writing all your own code anyway...

Like you can version the alias. That's a good way.

You can have an init that validates the content doesn't include any of the invalid states and converts them or removes them.
